src/scib_metrics/utils/_kmeans.py

```diff
@@ -1,21 +1,57 @@
 from functools import partial
+from typing import Literal
 
 import jax
 import jax.numpy as jnp
 import numpy as np
 from sklearn.utils import check_array
 
+from .._types import IntOrKey
 from ._dist import cdist
+from ._utils import validate_seed
 
 
-def _initialize_random(X: jnp.ndarray, n_clusters: int, key: jnp.ndarray) -> jnp.ndarray:
+def _initialize_random(X: jnp.ndarray, n_clusters: int, key: jax.random.KeyArray) -> jnp.ndarray:
     """Initialize cluster centroids randomly."""
     n_obs = X.shape[0]
     indices = jax.random.choice(key, n_obs, (n_clusters,), replace=False)
     initial_state = X[indices]
     return initial_state
 
 
+@partial(jax.jit, static_argnums=1)
+def _initialize_plus_plus(X: jnp.ndarray, n_clusters: int, key: jax.random.KeyArray) -> jnp.ndarray:
+    """Initialize cluster centroids with k-means++ algorithm."""
+
+    def _init(key):
+        key, subkey = jax.random.split(key)
+        # sample first centroid uniformly at random
+        idx = jax.random.choice(subkey, n_obs)
+        centroids = jnp.full((n_clusters,), -1, dtype=jnp.int32).at[0].set(idx)
+        mask = jnp.zeros((n_obs,), dtype=jnp.bool_).at[idx].set(True)
+        return centroids, mask, key
+
+    def _step(state):
+        centroids, mask, key = state
+        key, subkey = jax.random.split(key)
+        # d(x) = min_{mu in centroids} ||x - mu||^2, d(x) = 0 if x in centroids
+        probs = jnp.where(mask, 0, jnp.min(jnp.where(mask, dists, jnp.inf), axis=1) ** 2)
+        # sample with probability ~ d(x)
+        idx = jax.random.choice(subkey, n_obs, p=probs / jnp.sum(probs))
+        centroids = centroids.at[jnp.sum(mask)].set(idx)
+        mask = mask.at[idx].set(True)
+        return centroids, mask, key
+
+    def _convergence(state):
+        _, mask, _ = state
+        return jnp.sum(mask) < n_clusters
+
+    dists = cdist(X, X)
```

Comment:
why doesnt this get passed in? Kind of confusing and can lead to bugs potentially

Reply:
Hmm, don't the inner functions have access to outer variables? I was thinking it would be fine since `dists` is not being written to.

Reply:
yea it all technically works, its just a matter of code style to prevent bugs. I would recommend avoiding using variables that are instantiated later in the file, and also giving code the opportunity to mutate a shared variable. this is a just a nit, dw about changing it now.

Reply:
makes sense! thx for the feedback, I'll change it when I can